File: frontend/src/types/document.ts
/**
 * Document types for the flyer parser and document management
 */

export type DocumentType =
  | 'leasing_flyer'
  | 'void_analysis'
  | 'investment_memo'
  | 'loan_document'
  | 'comp_report'
  | 'other';

export type DocumentStatus = 'pending' | 'processing' | 'completed' | 'failed';

export interface AvailableSpace {
  id: string;
  document_id: string;
  property_id: string | null;
  suite_number: string | null;
  building_address: string | null;
  square_footage: number | null;
  min_divisible_sf: number | null;
  max_contiguous_sf: number | null;
  asking_rent_psf: number | null;
  rent_type: string | null;
  is_endcap: boolean;
  is_anchor: boolean;
  has_drive_thru: boolean;
  has_patio: boolean;
  features: Record<string, unknown> | null;
  notes: string | null;
  previous_tenant: string | null;
  created_at: string;
}

export interface ExistingTenant {
  id: string;
  document_id: string;
  property_id: string | null;
  name: string;
  category: string | null;
  suite_number: string | null;
  square_footage: number | null;
  is_anchor: boolean;
  is_national: boolean;
  created_at: string;
}

export interface ParsedDocument {
  id: string;
  user_id: string;
  property_id: string | null;
  filename: string;
  file_size: number;
  mime_type: string;
  page_count: number | null;
  document_type: DocumentType;
  confidence_score: number | null;
  status: DocumentStatus;
  error_message: string | null;
  extracted_data: ExtractedData | null;
  created_at: string;
  processed_at: string | null;
}

export interface ParsedDocumentDetail extends ParsedDocument {
  available_spaces: AvailableSpace[];
  existing_tenants: ExistingTenant[];
}

export interface DocumentListResponse {
  items: ParsedDocument[];
  total: number;
  page: number;
  page_size: number;
}

export interface DocumentUploadResponse {
  id: string;
  filename: string;
  status: DocumentStatus;
  message: string;
}

// Extracted data structures
export interface ExtractedPropertyInfo {
  name: string | null;
  address: string | null;
  city: string | null;
  state: string | null;
  zip_code: string | null;
  property_type: string | null;
  total_sf: number | null;
  year_built: number | null;
  parking_ratio: string | null;
  landlord_name: string | null;
}

export interface ExtractedAvailableSpace {
  suite_number: string | null;
  building_address: string | null;
  square_footage: number | null;
  min_divisible_sf: number | null;
  asking_rent_psf: number | null;
  rent_type: string | null;
  is_endcap: boolean;
  is_anchor: boolean;
  has_drive_thru: boolean;
  has_patio: boolean;
  previous_tenant: string | null;
  notes: string | null;
}

export interface ExtractedTenant {
  name: string;
  category: string | null;
  suite_number: string | null;
  square_footage: number | null;
  is_anchor: boolean;
  is_national: boolean;
}

export interface ExtractedFlyerData {
  property_info: ExtractedPropertyInfo;
  available_spaces: ExtractedAvailableSpace[];
  existing_tenants: ExtractedTenant[];
  amenities: string[];
  highlights: string[];
  contact_info: {
    broker_name: string | null;
    company: string | null;
    phone: string | null;
    email: string | null;
  } | null;
}

export interface VoidCategory {
  category_name: string;
  subcategory: string | null;
  is_void: boolean;
  site_count: number;
  market_count: number;
  avg_square_footage: number | null;
  existing_retailers: string[];
  void_opportunities: string[];
  match_score: number | null;
  common_cotenants: string[];
  notes: string | null;
}

export interface ExtractedVoidData {
  property_address: string | null;
  radius_miles: number | null;
  analysis_date: string | null;
  categories: VoidCategory[];
  summary: {
    total_categories_analyzed: number;
    total_voids: number;
    high_priority_voids: string[];
    key_insights: string[];
  };
}

export interface ExtractedInvestmentData {
  property_info: ExtractedPropertyInfo & {
    gla_sf: number | null;
    land_area_sf: number | null;
    status: string | null;
  };
  financials: {
    irr: number | null;
    rental_yield: number | null;
    exit_cap_rate: number | null;
    land_price: number | null;
    land_price_psf: number | null;
    total_investment: number | null;
    noi: number | null;
    asking_rent_psf: number | null;
    rent_type: string | null;
    cam_charges_psf: number | null;
  };
  demographics: {
    radius_miles: number | null;
    population: number | null;
    households: number | null;
    median_hh_income: number | null;
    avg_hh_income: number | null;
    daytime_employment: number | null;
    traffic_count: number | null;
  };
  tenant_interest: Array<{
    tenant_name: string;
    category: string | null;
    status: string | null;
    square_footage: number | null;
  }>;
  highlights: string[];
  scope_of_work: string | null;
  timing: {
    delivery_date: string | null;
    construction_start: string | null;
    lease_up_period: string | null;
  } | null;
}

// Union type for extracted data based on document type
export type ExtractedData = ExtractedFlyerData | ExtractedVoidData | ExtractedInvestmentData | Record<string, unknown>;

File: backend/app/models/document.py
"""
Pydantic models for document parsing API.
"""
from datetime import datetime
from enum import Enum
from uuid import UUID

from pydantic import BaseModel, Field


class DocumentType(str, Enum):
    LEASING_FLYER = "leasing_flyer"
    SITE_PLAN = "site_plan"
    VOID_ANALYSIS = "void_analysis"
    INVESTMENT_MEMO = "investment_memo"
    OFFERING_MEMORANDUM = "offering_memorandum"
    LOI = "loi"
    LOAN_DOCUMENT = "loan_document"
    COMP_REPORT = "comp_report"
    OTHER = "other"


class DocumentStatus(str, Enum):
    PENDING = "pending"
    PROCESSING = "processing"
    COMPLETED = "completed"
    FAILED = "failed"


# Available Space Models
class AvailableSpaceBase(BaseModel):
    suite_number: str | None = None
    building_address: str | None = None
    square_footage: int | None = None
    min_divisible_sf: int | None = None
    max_contiguous_sf: int | None = None
    asking_rent_psf: float | None = None
    rent_type: str | None = None
    is_endcap: bool = False
    is_anchor: bool = False
    has_drive_thru: bool = False
    has_patio: bool = False
    features: dict | None = None
    notes: str | None = None
    previous_tenant: str | None = None


class AvailableSpaceResponse(AvailableSpaceBase):
    id: UUID
    document_id: UUID
    property_id: UUID | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


# Existing Tenant Models
class ExistingTenantBase(BaseModel):
    name: str
    category: str | None = None
    suite_number: str | None = None
    square_footage: int | None = None
    is_anchor: bool = False
    is_national: bool = False


class ExistingTenantResponse(ExistingTenantBase):
    id: UUID
    document_id: UUID
    property_id: UUID | None = None
    created_at: datetime

    model_config = {"from_attributes": True}


# Parsed Document Models
class ParsedDocumentBase(BaseModel):
    filename: str
    document_type: DocumentType = DocumentType.OTHER


class ParsedDocumentResponse(BaseModel):
    id: UUID
    user_id: UUID
    property_id: UUID | None = None
    project_id: UUID | None = None
    filename: str
    file_size: int
    mime_type: str
    page_count: int | None = None
    document_type: DocumentType
    confidence_score: float | None = None
    status: DocumentStatus
    error_message: str | None = None
    extracted_data: dict | None = None
    is_archived: bool = False
    created_at: datetime
    processed_at: datetime | None = None

    model_config = {"from_attributes": True}


class ParsedDocumentDetailResponse(ParsedDocumentResponse):
    available_spaces: list[AvailableSpaceResponse] = []
    existing_tenants: list[ExistingTenantResponse] = []


class DocumentListResponse(BaseModel):
    items: list[ParsedDocumentResponse]
    total: int
    page: int
    page_size: int


# Document Upload Response
class DocumentUploadResponse(BaseModel):
    id: UUID
    filename: str
    status: DocumentStatus
    message: str


# Extracted Flyer Data (intermediate representation)
class ExtractedPropertyInfo(BaseModel):
    """Property information extracted from a leasing flyer."""
    name: str | None = None
    address: str | None = None
    city: str | None = None
    state: str | None = None
    zip_code: str | None = None
    property_type: str | None = None
    total_sf: int | None = None
    year_built: int | None = None
    parking_ratio: str | None = None
    landlord_name: str | None = None


class ExtractedAvailableSpace(BaseModel):
    """Available space extracted from a leasing flyer."""
    suite_number: str | None = None
    building_address: str | None = None
    square_footage: int | None = None
    min_divisible_sf: int | None = None
    asking_rent_psf: float | None = None
    rent_type: str | None = None
    is_endcap: bool = False
    is_anchor: bool = False
    has_drive_thru: bool = False
    has_patio: bool = False
    previous_tenant: str | None = None
    notes: str | None = None


class ExtractedTenant(BaseModel):
    """Existing tenant extracted from a leasing flyer."""
    name: str
    category: str | None = None
    suite_number: str | None = None
    square_footage: int | None = None
    is_anchor: bool = False
    is_national: bool = False


class ExtractedFlyerData(BaseModel):
    """Complete data extracted from a leasing flyer."""
    property_info: ExtractedPropertyInfo
    available_spaces: list[ExtractedAvailableSpace] = []
    existing_tenants: list[ExtractedTenant] = []
    amenities: list[str] = []
    highlights: list[str] = []
    contact_info: dict | None = None


# Site Plan Extracted Data Models
class SitePlanLocationArea(BaseModel):
    """A distinct area/zone within a site plan (building, pad, outparcel)."""
    name: str | None = None
    area_type: str | None = None  # building, pad, outparcel, parking
    square_footage: int | None = None
    tenant_name: str | None = None  # If occupied
    is_available: bool = False
    position_description: str | None = None  # e.g., "northwest corner", "inline"
    notes: str | None = None


class ExtractedSitePlanData(BaseModel):
    """Complete data extracted from a site plan/plot PDF."""
    property_info: ExtractedPropertyInfo
    location_areas: list[SitePlanLocationArea] = []
    tenant_locations: list[ExtractedTenant] = []  # Tenants with positions
    available_areas: list[SitePlanLocationArea] = []  # Vacant/available spaces
    total_site_sf: int | None = None
    parking_spaces: int | None = None
    parking_ratio: str | None = None
    site_dimensions: dict | None = None  # width, depth, etc.
    highlights: list[str] = []


# Start Analysis Response
class StartAnalysisResponse(BaseModel):
    """Response when starting void analysis from a document."""
    session_id: str
    property_address: str | None = None
    property_name: str | None = None
    tenant_count: int = 0
    available_space_count: int = 0
    document_type: DocumentType
    message: str = "Analysis session created successfully"


# Void Analysis Models
class VoidCategory(BaseModel):
    """A category in void analysis (e.g., Fast Casual, Boutique Fitness)."""
    category_name: str
    subcategory: str | None = None
    is_void: bool = True  # True if missing from trade area
    existing_count: int = 0
    market_count: int = 0
    match_score: float | None = None  # 0-100
    potential_tenants: list[str] = []
    notes: str | None = None


class VoidAnalysisCreate(BaseModel):
    property_id: str | None = None
    radius_miles: float = 3.0


class VoidAnalysisResponse(BaseModel):
    id: UUID
    document_id: UUID | None = None
    property_id: UUID | None = None
    user_id: UUID
    radius_miles: float | None = None
    analysis_date: datetime
    results: dict
    total_voids: int
    high_priority_voids: int
    created_at: datetime

    model_config = {"from_attributes": True}


# Investment Memo Models
class InvestmentMemoCreate(BaseModel):
    property_id: str | None = None
    title: str
    include_demographics: bool = True
    include_traffic: bool = True
    include_voids: bool = True
    include_tenant_interest: bool = True


class InvestmentMemoFinancials(BaseModel):
    irr: float | None = None
    rental_yield: float | None = None
    exit_cap_rate: float | None = None
    land_price: float | None = None
    total_investment: float | None = None
    noi: float | None = None
    asking_rent_psf: float | None = None
    rent_type: str | None = None


class InvestmentMemoResponse(BaseModel):
    id: UUID
    user_id: UUID
    property_id: UUID | None = None
    title: str
    version: int
    summary: str | None = None
    location_highlights: dict | None = None
    financials: dict | None = None
    demographics: dict | None = None
    tenant_interest: dict | None = None
    scope_of_work: str | None = None
    pdf_path: str | None = None
    is_draft: bool
    shared_with: list[str] | None = None
    created_at: datetime
    updated_at: datetime

    model_config = {"from_attributes": True}


class InvestmentMemoUpdate(BaseModel):
    title: str | None = None
    summary: str | None = None
    location_highlights: dict | None = None
    financials: dict | None = None
    demographics: dict | None = None
    tenant_interest: dict | None = None
    scope_of_work: str | None = None
    is_draft: bool | None = None
